### DISCUSSION TITLE
How can I use Detox to test camera, filepicker, and permissions for React Native application?

### DISCUSSION THREAD

**Discussion by anandb7248** [wix/Detox#3390] [Q&A]
Hello, is there any good documentation on how to test native features such as Camera, FilePicker, Permissions for a React Native application for both iOS and Android?

**Comment by portyu9** [comment]
Hi @anandb7248 

There isn't one Detox API that gives cross-platform control over camera UI, file pickers, and runtime permissions.

For **iOS simulator permissions**, Detox supports launch-time permission configuration:

```js
await device.launchApp({
  permissions: {
    camera: 'YES',
    photos: 'YES',
    microphone: 'YES',
  },
})
```

That Detox `permissions` option is iOS-only.

Android therefore needs a platform-specific permission strategy, such as preparing permissions before the test or automating the Android system permission UI when that dialog itself is what you want to test.

Camera and document/photo pickers are another boundary: once your application opens system UI or another application, you are no longer interacting purely with your React Native application's view hierarchy.

For stable E2E tests, I would normally put media/file selection behind an application abstraction:

```text
app code
 -> media/file abstraction
 -> real native implementation in production
 -> controlled implementation for E2E
```

Then Detox verifies that the expected file/media is selected and that your application responds correctly.

If your requirement is specifically to test the actual OS camera or picker UI, that part will need platform-specific automation rather than one identical Detox flow on Android and iOS.

Please mark this answer as accepted if it helped, thank you.
